go/types/validating_decoder.go

```diff
@@ -0,0 +1,40 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package types
+
+import (
+	"github.com/attic-labs/noms/go/chunks"
+	"github.com/attic-labs/noms/go/d"
+)
+
+type ValidatingDecoder struct {
+	vs *ValueStore
+}
+
+func NewValidatingDecoder(cs chunks.ChunkStore) *ValidatingDecoder {
+	return &ValidatingDecoder{
+		newLocalValueStore(cs),
+	}
+}
+
+// DecodedChunk holds a pointer to a Chunk and the Value that results from
+// calling DecodeFromBytes(c.Data()).
+type DecodedChunk struct {
+	Chunk *chunks.Chunk
+	Value *Value
+}
+
+// Decode decodes c and checks that the hash of the resulting value
+// matches c.Hash(). It returns a DecodedChunk holding both c and a pointer to
+// the decoded Value.
+func (vbs *ValidatingDecoder) Decode(c *chunks.Chunk) DecodedChunk {
+	h := c.Hash()
+	v := decodeFromBytesWithValidation(c.Data(), vbs.vs)
```

Comment:
I think it would probably be OK for the current usage of the API, but since this is API is visible outside the package I can't guarantee that all future callers will know what's safe to do with the returned Value and what's not. So, I erred on the side of safety and least-surprise.